File: ChineseCheckers/app/src/main/java/com/chinesecheckers/game/GameLogic.kt
package com.chinesecheckers.game

import kotlin.math.abs

object GameLogic {

    fun handleTap(state: GameState, pos: BoardPos): GameState {
        if (state.winner != null) return state

        val pieceAtPos = state.pieces[pos]

        return when {
            // Mid-jump: tapping current piece position ends the jump chain
            state.isJumping && pos == state.jumpPos -> endJump(state)

            // Mid-jump: only valid jump continuations accepted
            state.isJumping && pos in state.validMoves -> executeMove(state, state.jumpPos!!, pos)

            // Mid-jump: ignore everything else
            state.isJumping -> state

            // Select or reselect own piece
            pieceAtPos == state.currentPlayer -> selectPiece(state, pos)

            // Move selected piece to valid destination
            state.selectedPos != null && pos in state.validMoves -> executeMove(state, state.selectedPos, pos)

            // Deselect by tapping elsewhere
            state.selectedPos != null -> state.copy(selectedPos = null, validMoves = emptySet())

            else -> state
        }
    }

    fun endJump(state: GameState): GameState {
        if (!state.isJumping) return state
        val next = if (state.currentPlayer == 1) 2 else 1
        return state.copy(
            currentPlayer = next,
            selectedPos = null,
            validMoves = emptySet(),
            isJumping = false,
            jumpPos = null,
            jumpVisited = emptySet()
        )
    }

    private fun selectPiece(state: GameState, pos: BoardPos): GameState {
        val moves = computeValidMoves(pos, state.pieces, emptySet(), isTurnStart = true)
        return state.copy(selectedPos = pos, validMoves = moves)
    }

    private fun executeMove(state: GameState, from: BoardPos, to: BoardPos): GameState {
        val newPieces = state.pieces.toMutableMap()
        val player = newPieces.remove(from)!!
        newPieces[to] = player

        val jumped = isJump(from, to)

        if (jumped) {
            val visited = state.jumpVisited + from + to
            val moreJumps = computeJumpMoves(to, newPieces, visited)
            if (moreJumps.isNotEmpty()) {
                return state.copy(
                    pieces = newPieces,
                    isJumping = true,
                    jumpPos = to,
                    jumpVisited = visited,
                    validMoves = moreJumps,
                    selectedPos = to
                )
            }
        }

        val winner = checkWinner(newPieces)
        val next = if (state.currentPlayer == 1) 2 else 1
        return state.copy(
            pieces = newPieces,
            currentPlayer = next,
            selectedPos = null,
            validMoves = emptySet(),
            isJumping = false,
            jumpPos = null,
            jumpVisited = emptySet(),
            winner = winner
        )
    }

    private fun computeValidMoves(
        from: BoardPos,
        pieces: Map<BoardPos, Int>,
        visited: Set<BoardPos>,
        isTurnStart: Boolean
    ): Set<BoardPos> {
        val moves = mutableSetOf<BoardPos>()

        if (isTurnStart) {
            // Single-step moves (only at start of turn)
            for (n in Board.getNeighbors(from)) {
                if (n !in pieces) moves += n
            }
        }

        // Jump moves (recursive)
        moves += computeJumpMoves(from, pieces, visited + from)
        return moves
    }

    private fun computeJumpMoves(
        from: BoardPos,
        pieces: Map<BoardPos, Int>,
        visited: Set<BoardPos>
    ): Set<BoardPos> {
        val moves = mutableSetOf<BoardPos>()
        for (n in Board.getNeighbors(from)) {
            if (n in pieces) {
                val target = Board.jumpTarget(from, n)
                if (target in Board.allPositions && target !in pieces && target !in visited) {
                    moves += target
                    moves += computeJumpMoves(target, pieces, visited + target)
                }
            }
        }
        return moves
    }

    private fun isJump(from: BoardPos, to: BoardPos): Boolean =
        abs(from.row - to.row) > 1 || abs(from.col - to.col) > 2

    private fun checkWinner(pieces: Map<BoardPos, Int>): Int? {
        val p1 = pieces.filter { it.value == 1 }.keys
        val p2 = pieces.filter { it.value == 2 }.keys
        if (p1.isNotEmpty() && p1.all { it in Board.player2Home }) return 1
        if (p2.isNotEmpty() && p2.all { it in Board.player1Home }) return 2
        return null
    }
}

File: ChineseCheckersIOS/ChineseCheckersIOS/Game/GameLogic.swift
import Foundation

enum GameLogic {

    static func handleTap(_ state: GameState, pos: BoardPos) -> GameState {
        guard state.winner == nil else { return state }
        // Once hasMoved and not in a jump chain, only Confirm/Undo work
        if state.hasMoved && !state.isJumping { return state }

        if state.isJumping {
            if state.validMoves.contains(pos) { return execute(state, from: state.jumpPos!, to: pos) }
            return state
        }

        let pieceHere = state.pieces[pos]
        if let p = pieceHere, p == state.currentPlayer { return selectPiece(state, pos: pos) }
        if state.selectedPos != nil && state.validMoves.contains(pos) { return execute(state, from: state.selectedPos!, to: pos) }
        if state.selectedPos != nil {
            var s = state; s.selectedPos = nil; s.validMoves = []; return s
        }
        return state
    }

    static func confirmTurn(_ state: GameState) -> GameState {
        var s = state
        s.currentPlayer = state.currentPlayer == 1 ? 2 : 1
        s.selectedPos = nil; s.validMoves = []
        s.isJumping = false; s.jumpPos = nil; s.jumpVisited = []
        s.hasMoved = false
        s.moveCount += 1
        return s
    }

    // MARK: - Private

    private static func selectPiece(_ state: GameState, pos: BoardPos) -> GameState {
        var s = state
        s.selectedPos = pos
        s.validMoves = validMoves(from: pos, pieces: state.pieces, visited: [pos])
        return s
    }

    private static func execute(_ state: GameState, from: BoardPos, to: BoardPos) -> GameState {
        var pieces = state.pieces
        let player = pieces.removeValue(forKey: from)!
        pieces[to] = player

        let winner = checkWinner(pieces: pieces)
        let jumped = isJump(from: from, to: to)

        var s = state
        s.pieces = pieces
        s.hasMoved = true
        s.winner = winner

        if jumped && winner == nil {
            let visited = state.jumpVisited.union([from, to])
            let more = jumpMoves(from: to, pieces: pieces, visited: visited)
            if !more.isEmpty {
                s.isJumping = true; s.jumpPos = to
                s.jumpVisited = visited; s.validMoves = more; s.selectedPos = to
                return s
            }
        }

        s.isJumping = false; s.jumpPos = nil; s.jumpVisited = []
        s.validMoves = []; s.selectedPos = to
        return s
    }

    private static func validMoves(from: BoardPos, pieces: [BoardPos: Int], visited: Set<BoardPos>) -> Set<BoardPos> {
        var moves = Set<BoardPos>()
        for n in Board.neighbors(of: from) where pieces[n] == nil { moves.insert(n) }
        moves.formUnion(jumpMoves(from: from, pieces: pieces, visited: visited))
        return moves
    }

    private static func jumpMoves(from: BoardPos, pieces: [BoardPos: Int], visited: Set<BoardPos>) -> Set<BoardPos> {
        var moves = Set<BoardPos>()
        let directions: [(Int, Int)] = [(0, 2), (0, -2), (1, 1), (1, -1), (-1, 1), (-1, -1)]
        for (dr, dc) in directions {
            var k = 1
            while true {
                let scanned = BoardPos(row: from.row + k * dr, col: from.col + k * dc)
                guard Board.allPositions.contains(scanned) else { break }
                if pieces[scanned] != nil {
                    let target = BoardPos(row: from.row + 2 * k * dr, col: from.col + 2 * k * dc)
                    if Board.allPositions.contains(target) && pieces[target] == nil && !visited.contains(target) {
                        moves.insert(target)
                        moves.formUnion(jumpMoves(from: target, pieces: pieces, visited: visited.union([target])))
                    }
                    break
                }
                k += 1
            }
        }
        return moves
    }

    private static func isJump(from: BoardPos, to: BoardPos) -> Bool {
        abs(from.row - to.row) > 1 || abs(from.col - to.col) > 2
    }

    private static func checkWinner(pieces: [BoardPos: Int]) -> Int? {
        let p1 = pieces.filter { $0.value == 1 }.keys
        let p2 = pieces.filter { $0.value == 2 }.keys
        if !p1.isEmpty && p1.allSatisfy({ Board.player2Home.contains($0) }) { return 1 }
        if !p2.isEmpty && p2.allSatisfy({ Board.player1Home.contains($0) }) { return 2 }
        return nil
    }
}
